File: LeetCode/Python Solutions/Greatest Common Divisor Traversal/greatest.py
from typing import List
from collections import defaultdict

class Solution:
    def canTraverseAllPairs(self, nums: List[int]) -> bool:
        n = len(nums)
        prime_to_indices = defaultdict(list)

        # Function to factorize a number and update prime_to_indices mapping
        def factorize(num):
            factors = set()
            d = 2
            while d * d <= num:
                while num % d == 0:
                    factors.add(d)
                    num //= d
                d += 1
            if num > 1:
                factors.add(num)
            return factors
        
        # Populate prime_to_indices with factors of each number
        for i, num in enumerate(nums):
            for factor in factorize(num):
                prime_to_indices[factor].append(i)
        
        # Union-Find data structure
        parent = list(range(n))
        
        def find(x):
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]
        
        def union(x, y):
            parent[find(x)] = find(y)
        
        # Union indices with common prime factors
        for indices in prime_to_indices.values():
            for i in range(1, len(indices)):
                union(indices[i], indices[i-1])
        
        # Check if all numbers are connected
        root = find(0)
        for i in range(1, n):
            if find(i) != root:
                return False
        return True

File: LeetCode/Java Solutions/Greatest Common Divisor Traversal/greatest.java
import java.util.*;

public class Solution {
    // Union-Find class for managing disjoint sets
    class UnionFind {
        private int[] parent;

        public UnionFind(int size) {
            parent = new int[size];
            for (int i = 0; i < size; i++) {
                parent[i] = i;
            }
        }

        public int find(int x) {
            if (parent[x] != x) {
                parent[x] = find(parent[x]);
            }
            return parent[x];
        }

        public void union(int x, int y) {
            parent[find(x)] = find(y);
        }
    }

    public boolean canTraverseAllPairs(int[] nums) {
        int n = nums.length;
        Map<Integer, List<Integer>> primeToIndices = new HashMap<>();
        UnionFind uf = new UnionFind(n);

        for (int i = 0; i < n; i++) {
            for (int factor : factorize(nums[i])) {
                primeToIndices.computeIfAbsent(factor, k -> new ArrayList<>()).add(i);
            }
        }

        for (List<Integer> indices : primeToIndices.values()) {
            for (int i = 1; i < indices.size(); i++) {
                uf.union(indices.get(0), indices.get(i));
            }
        }

        int root = uf.find(0);
        for (int i = 1; i < n; i++) {
            if (uf.find(i) != root) {
                return false;
            }
        }
        return true;
    }

    private List<Integer> factorize(int num) {
        Set<Integer> factors = new HashSet<>();
        for (int factor = 2; factor * factor <= num; factor++) {
            while (num % factor == 0) {
                factors.add(factor);
                num /= factor;
            }
        }
        if (num > 1) {
            factors.add(num);
        }
        return new ArrayList<>(factors);
    }
}
